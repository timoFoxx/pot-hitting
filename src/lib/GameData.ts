export enum Dimensions {
    height = 600,
    width = 500,
}

export enum PotCoords {
    potX = Math.random() * Dimensions.height,
    potY = Math.random() * Dimensions.width,
}
