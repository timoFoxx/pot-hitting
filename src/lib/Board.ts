import { Point } from "./Game";


export type potX = number;

export type potY = number;

export class Board {
    // Upper left corner point is denoted by x and y
    potWidth: number = 20
    potHeight: number = 20

    constructor(public x: potX, public y: potY) {}

    isOnPot = (p: Point): Boolean => {
        return Math.abs(p.x - this.x) < this.potWidth && Math.abs(p.y - this.y) < this.potHeight
    }

}
