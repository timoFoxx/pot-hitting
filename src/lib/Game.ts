import { Board } from './Board'

export class Point {
    constructor(public x: number, public y: number) {}

    //calculates distance from mouse pointer to pot coords
    distanceFrom = (p: Point): number  => Math.sqrt(Math.pow(this.x - p.x, 2) + Math.pow(this.y - p.y, 2))
}

export interface GameEvent {
    score: number;
}

export class Win implements GameEvent {
    constructor(public score: number) {}
}

export class Cold implements GameEvent {
    constructor(public score: number) {}
}

export class Warm implements GameEvent {
    constructor(public score: number) {}
}

export class FirstClick implements GameEvent {
    constructor(public score: number) {}
}

//side effects
export class GameState {
    constructor(public distance: number | null, public score: number) {}

    setDistance = (p: number) => {
        this.distance = p
    }

    incrementScore = () => {
        this.score += 1;
    }
}

export type GameLoop = (e: Point) => GameEvent

export const Game = (board: Board, state: GameState = new GameState(null, 0)): GameLoop => (userEvent: Point): GameEvent => {
    const newDistance = userEvent.distanceFrom(new Point(board.x, board.y))
    // side effects here -> redux?
    if (board.isOnPot(userEvent)) {
        state.incrementScore()
        return new Win(state.score)
    } else if (state.distance == null) {
        state.setDistance(newDistance)
        return new FirstClick(state.score)
    } else if (state.distance > newDistance) {
        state.setDistance(newDistance)
        state.incrementScore()
        return new Warm(state.score)
    } else {
        state.setDistance(newDistance)
        state.incrementScore()
        return new Cold(state.score)
    }
}
