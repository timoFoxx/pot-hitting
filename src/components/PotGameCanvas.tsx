import React, { useEffect, useState } from 'react'
import { useRef } from 'react'
import { StyledPotGame } from "../styles/Layout";
import {PotCoords} from "../lib/GameData";

import {GameEvent, Point, Warm} from '../lib/Game'

interface PotGameCanvasProps {
    width: number
    height: number
    onClick: (p: Point) => void
    gameEvent: GameEvent | null
}


export const PotGameCanvas = (props: PotGameCanvasProps) => {
    const canvasRef = useRef<HTMLCanvasElement>(null);

    const getCoordinates = (event: MouseEvent): Point  => {


        if (canvasRef.current) {
            const canvas: HTMLCanvasElement = canvasRef.current;
            return new Point(event.pageX - canvas.offsetLeft, event.pageY - canvas.offsetTop)
        } return new Point(event.x, event.y);
    };



    useEffect(() => {
        if (!canvasRef.current) {
            return;
        }

        canvasRef.current.addEventListener('click', (event: MouseEvent) => {

            props.onClick(getCoordinates(event));
            if (canvasRef.current) {
                if (props.gameEvent instanceof Warm) {
                    const context = canvasRef.current.getContext('2d')
                    if (context) {
                        context.fillStyle = '#000000'
                        context.fillRect(PotCoords.potX, PotCoords.potY, 20, 20);

                        canvasRef.current.removeEventListener('click', (event: MouseEvent) => {
                            props.onClick(getCoordinates(event));
                        }, false);

                    }
                }
            }

        })

    },[])


    return <StyledPotGame ref={canvasRef} height={props.height} width={props.width} />;
}



