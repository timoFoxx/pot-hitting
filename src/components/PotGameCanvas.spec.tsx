import React from 'react'
import renderer from 'react-test-renderer'
import {Dimensions} from "../lib/GameData";
import { Win } from '../lib/Game'

import { PotGameCanvas} from "./PotGameCanvas";

const onClick = jest.fn();

describe('PotGameCanvas', () => {
    test('should render according to the snapshot when created', () => {
        expect(renderer.create(<PotGameCanvas width={Dimensions.width} height={Dimensions.height} onClick={onClick} gameEvent={new Win(20)} />).toJSON()).toMatchSnapshot()
    })

})
