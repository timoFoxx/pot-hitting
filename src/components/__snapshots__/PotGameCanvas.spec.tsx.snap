// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PotGameCanvas should render according to the snapshot when created 1`] = `
<canvas
  className="sc-dlnjPT cjkAhu"
  height={600}
  width={500}
/>
`;

exports[`PotGameCanvas should render according to the snapshot when won 1`] = `
<canvas
  className="sc-dlnjPT cjkAhu"
  height={600}
  width={500}
/>
`;
